motionplan/tpSpaceRRT.go

```diff
@@ -172,185 +205,329 @@ func (mp *tpSpaceRRTMotionPlanner) planRunner(
 		}
 	}
 
-	dist := math.Sqrt(mp.planOpts.DistanceFunc(&Segment{StartPosition: startPose, EndPosition: goalPose}))
-	midPt := goalPose.Point().Sub(startPose.Point())
+	m1chan := make(chan *nodeAndError, 1)
+	m2chan := make(chan *nodeAndError, 1)
+	defer close(m1chan)
+	defer close(m2chan)
+
+	dist := math.Sqrt(mp.planOpts.DistanceFunc(&ik.Segment{StartPosition: startPose, EndPosition: goalPose}))
+	midptNode := &basicNode{pose: spatialmath.Interpolate(startPose, goalPose, 0.5), cost: dist} // used for initial seed
+	var randPosNode node = midptNode
 
-	var randPos spatialmath.Pose
 	for iter := 0; iter < mp.planOpts.PlanIter; iter++ {
 		if ctx.Err() != nil {
 			mp.logger.Debugf("TP Space RRT timed out after %d iterations", iter)
 			rrt.solutionChan <- &rrtPlanReturn{planerr: fmt.Errorf("TP Space RRT timeout %w", ctx.Err()), maps: rrt.maps}
 			return
 		}
-		// Get random cartesian configuration
-		tryGoal := true
-		// Check if we can reach the goal every N iters
-		if iter%mp.algOpts.goalCheck != 0 {
-			rDist := dist * (mp.algOpts.autoBB + float64(iter)/10.)
-			tryGoal = false
-			randPosX := float64(mp.randseed.Intn(int(rDist)))
-			randPosY := float64(mp.randseed.Intn(int(rDist)))
-			randPosTheta := math.Pi * (mp.randseed.Float64() - 0.5)
-			randPos = spatialmath.NewPose(
-				r3.Vector{midPt.X + (randPosX - rDist/2.), midPt.Y + (randPosY - rDist/2.), 0},
-				&spatialmath.OrientationVector{OZ: 1, Theta: randPosTheta},
-			)
-		} else {
-			randPos = goalPose
+
+		utils.PanicCapturingGo(func() {
+			m1chan <- mp.attemptExtension(ctx, randPosNode, rrt.maps.startMap, false)
+		})
+		utils.PanicCapturingGo(func() {
+			m2chan <- mp.attemptExtension(ctx, randPosNode, rrt.maps.goalMap, true)
+		})
+		seedMapReached := <-m1chan
+		goalMapReached := <-m2chan
+
+		seedMapNode := seedMapReached.node
+		goalMapNode := goalMapReached.node
+		err := multierr.Combine(seedMapReached.error, goalMapReached.error)
+		if err != nil {
+			rrt.solutionChan <- &rrtPlanReturn{planerr: err, maps: rrt.maps}
+			return
 		}
-		randPosNode := &basicNode{q: nil, cost: 0, pose: randPos, corner: false}
 
-		successNode := mp.attemptExtension(ctx, nil, randPosNode, rrt)
-		// If we tried the goal and have a close-enough XY location, check if the node is good enough to be a final goal
-		if tryGoal && successNode != nil {
-			if mp.planOpts.GoalThreshold > mp.planOpts.goalMetric(&State{Position: successNode.Pose(), Frame: mp.frame}) {
-				// If we've reached the goal, break out
-				path := extractPath(rrt.maps.startMap, rrt.maps.goalMap, &nodePair{a: successNode})
+		if seedMapNode != nil && goalMapNode != nil {
+			seedReached := mp.attemptExtension(ctx, goalMapNode, rrt.maps.startMap, false)
+			if seedReached.error != nil {
+				rrt.solutionChan <- &rrtPlanReturn{planerr: seedReached.error, maps: rrt.maps}
+				return
+			}
+			if seedReached.node == nil {
```

Comment:
Doesn't look like we flip the maps anymore and that might be a good thing to do because as is this will bias the growth the the starting map over the goal map

Reply:
On the contrary, with some newer improvements to cbirrt, we don't really need to flip the maps anymore either. We do the same thing here as with cbirrt, that is, we extend both trees towards the random point, then extend each towards the other. The swapping of maps hasn't been necessary in cbirrt for some time now.